fetchai/ledger/crypto/deed.py

```diff
@@ -0,0 +1,66 @@
+# ------------------------------------------------------------------------------
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+import logging
+from typing import Union
+
+from fetchai.ledger.crypto import Address, Identity, Entity
+
+AddressLike = Union[Address, Identity]
+
+class Deed:
+    def __init__(self, address: Entity):
+        self._address = address
+        self._amend_threshold = None
+        self._transfer_threshold = None
+        self._signees = {}
+
+    def add_signee(self, signee: Entity, voting_weight: int):
+        self._signees[signee] = voting_weight
+
+    @property
+    def amend_threshold(self):
+        return self._amend_threshold
+
+    @amend_threshold.setter
+    def amend_threshold(self, value):
+        self._amend_threshold = value
+
+    @property
+    def transfer_threshold(self):
+        return self._transfer_threshold
+
+    @transfer_threshold.setter
+    def transfer_threshold(self, value):
+        self._transfer_threshold = value
```

Comment:
As by now, the ledger recognises multiple other deed operations on top of `amend` and `transfer`:
- `execute`
- `stake`

But, in general, deed allows to define/provide any number of operations with any name.

Reply:
Thus it would be perhaps better to make the interface based on string keys (that's what the deed operation type actually in reality is, e.g.:

```python
from enum import Enum, auto

class NoValue(Enum):
    def __repr__(self):
        return '<%s.%s>' % (self.__class__.__name__, self.name)
    def __str__(self):
        return '%s' % (self.name)

class Operation(NoValue):
    amend = auto()
    transfer = auto()
    execute = auto()
    stake = auto()
...

class Deed:
    ...
    def threshold(operation: Operation, value: int):
        deed['thresholds'][str(operation)] = value
    ...
```

Then usage:

```
d = Deed()
d.threshold(Operaion.stake, 1)
```

Reply:
I agree this is neater - have implemented and tested.